transforms/ember-object/index.js

```diff
@@ -0,0 +1,197 @@
+      return j.methodDefinition("method", functionProp.key, functionProp.value);
+    });
+  }
+
+  /**
+   * Create  a constructor method
+   *
+   * @param {[]} instanceProps
+   * @return Array [MethodDefinition]
+   */
+  function createConstructor(instanceProps = []) {
+    if (instanceProps.length) {
+      return [
+        j.methodDefinition(
+          "constructor",
+          j.identifier("constructor"),
+          j.functionExpression(
+            null,
+            [],
+            j.blockStatement(instancePropsToExpressions(instanceProps))
+          )
+        )
+      ]; // Todo: find a way to include params
+    }
+
+    return [];
+  }
+
+  /**
+   * Create the class
+   *
+   * @param {String} className
+   * @param {Object} { instanceProps, functionProps } map of instance and function properties
+   */
+  function createClass(className, { instanceProps, functionProps } = {}) {
+    return j.classDeclaration(
+      j.identifier(className),
+      j.classBody(
+        []
+          .concat(createConstructor(instanceProps))
+          .concat(functionPropsToMethods(functionProps))
+      ),
+      j.identifier("EmberObject")
+    );
+  }
+
+  /**
+   * Return the map of instance props and functions from Ember Object
+   *
+   * For example
+   * const myObj = EmberObject.extend({ key: value, foo: function() {}, baz() {} });
+   * will be parsed as:
+   *{
+      instanceProps: [ key: value ]
+      functionProps: [ FunctionExpression(foo), FunctionExpression(baz) ]
+    }
+   * @param {*} emberObjectExpression
+   * @returns { instanceProps, functionProps } map of instance and function properties
+   */
+  function getEmberObjectProperties(emberObjectExpression) {
+    const objProps = j(emberObjectExpression)
+      .find(j.ObjectExpression)
+      .get("properties").value;
+
+    const instanceProps = [];
+    const functionProps = [];
+
+    objProps.forEach(prop => {
+      if (prop.value.type === "FunctionExpression") {
+        functionProps.push(prop);
+      } else {
+        instanceProps.push(prop);
+      }
+    });
+
+    return { instanceProps, functionProps };
+  }
+
+  function getEmberObjectExpressions(root) {
+    return root.find(j.CallExpression, {
+      callee: {
+        object: {
+          name: "EmberObject"
+        },
+        property: {
+          name: "extend"
+        }
+      }
+    });
+  }
+
+  /**
+   * Parse input and find the variable declaration with value as EmberObject
+   *
+   * This code looks for pattern: const myObj = EmberObject.extend({});
+   * @param {jscodeshift root} root
+   * @returns Array [ VariableDeclarator ]
+   */
+  function getEmberObjectExtendVarDeclarator(root) {
+    return root.find(j.VariableDeclaration, {
+      declarations: [
+        {
+          type: "VariableDeclarator",
+          init: {
+            type: "CallExpression",
+            callee: {
+              object: {
+                name: "EmberObject"
```

Comment:
Based on this it looks like we will only be transforming `EmberObject.extend`, and no other classes (e.g. `Component.extend`, `Controller.extend`, etc). This is a good start, and definitely on the safer end of things, but I think we'll want to develop some more general heuristics.

We could try a whitelist, but that won't account for any user created classes which are extended. I _think_ we'd be pretty safe in general if the callee is CapitalCased, which conventionally means it's a class definition.

Reply:
Yes, this one is only targeted for `EmberObject`. Right now I am focused only on small objects which do not have any computed properties or any non standard stuff. I am going to add a validation logic so that we can make an early exit if the object have any of those properties.

We can definitely iterate over it and make this applicable to `Component` `Controllers` or any user generated objects using `extend`.

Reply:
Ya I think we should just focus on the simple cases for now, which likely will prove to be fairly complex in and of themselves. Once we do that, we can move to generalize this to handle more types.

Does this sound OK?

Reply:
If we're talking about the "stages" concept when we say simple, I think the issue is that what makes a case "simple" isn't whether or not the class extends from `EmberObject` directly, but what features it uses. A Component or Service can be simple as long as it has no computed properties or other decorator things, and a plain `EmberObject` class can be unconvertible.

That said, totally think it's reasonable to avoid adding the heuristic until we have other kinks worked out. We can iterate towards that.